lib/src/main/java/com/squareup/cycler/Update.kt

```diff
@@ -120,49 +166,68 @@ class Update<I : Any>(private val oldRecyclerData: RecyclerData<I>) {
   internal fun generateUpdateWork(itemComparator: ItemComparator<I>?): UpdateWork {
 
     val extraItemChanged = oldRecyclerData.extraItem != extraItem
-    val refreshAllNeeded = dataReplaced && itemComparator == null
 
     val asyncWork = mutableListOf<() -> Unit>()
     val notifications = mutableListOf<(Adapter<*>) -> Unit>()
 
-    if (refreshAllNeeded) {
-      @SuppressLint("NotifyDataSetChanged")
-      notifications += { adapter -> adapter.notifyDataSetChanged() }
-    } else {
+    if (extraItemChanged) {
+      // We notify first the extraItem change, so we use its position in oldRecyclerData.
+      notifications += notifyChangesExtraItem()
+    }
 
-      if (extraItemChanged) {
-        // We notify first the extraItem change, so we use its position in oldRecyclerData.
-        notifications += notifyChangesExtraItem()
+    when {
+      dataPopulated -> {
+        val work = generateUpdateLambdas(populateMode, itemComparator)
+        work.first?.let { asyncWork.add(it) }
+        notifications.add(work.second)
```

Comment:
if dataPopulated -> we just need to send notifyDataSetChanged, no diffing even if we have a comparator!
Just do `notifications.add { it.notifydataSetChanged() }`. If populateMode == async then add an empty lambda to asyncWork: `asyncWork += {}`